File: Day 2/AOC2023_2_P1.nim
import std/tables
import std/strutils
import std/sugar
import std/math

const max = {"red": 12, "green": 13, "blue": 14}.toTable

proc processLine(line: string) : int = 
    var ok = true
    let id = line.split(" ")[1].strip(chars = {':'})
    let infos = line.split(seps = {':', ',', ';'})[1 .. ^1]
    for elt in infos:
        let (v, k) = (elt.strip().split(" ")[0], elt.strip().split(" ")[1])
        if max[k] < parseInt(v):
            ok = false
    if ok:
        return parseInt(id)
    else:
        return 0

let res = collect:
    for elt in lines("inputs"):
        processLine(elt)
echo sum(res)

File: Day 2/AOC2023_2_P1.rb
MAX = { "red" => 12, "green" => 13, "blue" => 14}

def extractVal(set)
    set.split(",").map do |x|
        [x.split(" ")[1], x.split(" ")[0].to_i]
    end.to_h
end

def processLine(line)
    id = /Game ([0-9]+)/.match(line).captures[0]
    ok = true
    # Process each set in a game
    line.split(":")[-1].split(";").each do |x| 
        extractVal(x).each_pair do |k, v|
            if MAX[k] < v
                ok = false
            end
        end
    end
    return id if ok
end

puts File.open("inputs").each.map {|x| processLine(x).to_i}.select {|x| x != nil}.sum